### PR TITLE
MCR-2159 new prematching filter MCRNormalizeMCRObjectIDsFilter

### PR DESCRIPTION
**Pull request by rsteph-de**
[Link to jira](https://mycore.atlassian.net/browse/MCR-2159).

### CODE REVIEW

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java  18
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5289482&bid=17317485)

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 8ffe3c0b907aeef64ea906ebbfac6f34652a6291! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/5403-pull_request-949

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                        } catch (Exception e) {
+                            LOGGER.error("Error retrieving object id from SOLR", e);
+                        }
+                    }
+                } else {
+                    MCRObjectID mcrObjID = MCRObjectID.getInstance(mcrid);
+                    // set the properly formated mcrObjID back to URL
+                    mcrid = mcrObjID.toString();
+                    if (!mcrid.equals(pathParts[1])) {
+                        pathParts[1] = mcrid;
```
Comment (yagee-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java]: the check for equals is not needed here, as assigning an array element is far less complex than checking for equal

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+                        params.set("start", 0);
+                        params.set("rows", 1);
+                        params.set("fl", "id");
+                        params.set("fq", "objectKind:mycoreobject");
+                        params.set("q", key + ":" + ClientUtils.escapeQueryChars(value));
+                        QueryResponse solrResponse;
+                        try {
+                            solrResponse = MCRSolrClientFactory.getMainSolrClient().query(params);
+                            SolrDocumentList solrResults = solrResponse.getResults();
+                            if (solrResults.getNumFound() >= 1) {
```
Comment (yagee-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java]: should only work if getNumFound() == 1!
if getNumFound() > 1 return 400
if getNumFound() = 1 return 404

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+ * @author Robert Stephan
+ *
+ */
+@Provider
+@PreMatching
+public class MCRNormalizeMCRObjectIDsFilter implements ContainerRequestFilter {
+
+    private static final Logger LOGGER = LogManager.getLogger();
+
+    private static List<String> SEARCHKEYS_FOR_OBJECTS = Arrays.asList(
```
Comment (yagee-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java]: Isn't a HashSet better for searching?
```
MCRConfiguration2.getString("MCR.RestAPI.V2.AlternativeIdentifier.Objects.Keys")
    .stream()
    .flatMap(MCRConfiguration2::splitValue)
    .collect(Collectors.toSet());
```

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                    // set the properly formated mcrObjID back to URL
+                    mcrid = mcrObjID.toString();
+                    if (!mcrid.equals(pathParts[1])) {
+                        pathParts[1] = mcrid;
+                    }
+                }
+            } catch (MCRException ex) {
+                // ignore
+
+            }
```
Comment (yagee-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java]: externalize method: getMCRObjectId(String)

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            MCRConfiguration2.getString("MCR.RestAPI.V2.AlternativeIdentifier.Derivates.Keys").orElse("").split(","));
+
+    @Context
+    ResourceInfo resourceInfo;
+
+    @Context
+    HttpServletResponse response;
+
+    @Override
+    public void filter(ContainerRequestContext requestContext) throws IOException {
```
Comment (yagee-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java:76]: return early if SEARCHKEYS_FOR_* are empty

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                        MCRObjectID mcrDerID = MCRObjectID.getInstance(derid);
+                        // set the properly formated mcrObjID back to URL
+                        derid = mcrDerID.toString();
+                        if (!derid.equals(pathParts[3])) {
+                            pathParts[3] = derid;
+                        }
+                    }
+                } catch (MCRException ex) {
+                    // ignore
+                }
```
Comment (yagee-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java]: externalize method: getDerivateId(String)

**Review by yagee-de — CHANGES_REQUESTED** [review]

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java (hunk 1 above)
Comment (rsteph-de) [reply mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java]: O.K.

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java (hunk 2 above)
Comment (rsteph-de) [reply mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java]: O.K.

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java (hunk 3 above)
Comment (rsteph-de) [reply mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java]: O.K.

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java (hunk 5 above)
Comment (rsteph-de) [reply mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java:76]: O.K.

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java (hunk 6 above)
Comment (rsteph-de) [reply mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java]: O.K.

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java (hunk 4 above)
Comment (rsteph-de) [reply mycore-restapi/src/main/java/org/mycore/restapi/MCRNormalizeMCRObjectIDsFilter.java]: O.K.

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 962308679aaeb9e1fbab7b7198aae377da94795b! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/5419-pull_request-949

**Review by yagee-de — APPROVED** [review]

**Pull request merged**